src/mlpack/methods/ann/image_functions/bilinear_function.hpp

```diff
@@ -0,0 +1,143 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_INTERPOLATION_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_INTERPOLATION_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The bilinear interpolation function
+ *
+ */
+class BiLinearFunction
+{
+ public:
+  BiLinearFunction(
+      const size_t inRowSize,
+      const size_t inColSize,
+      const size_t scale):
+      inRowSize(inRowSize),
+      inColSize(inColSize),
+      outRowSize(2 * inRowSize),
+      outColSize(2 * inColSize),
+      scale(scale)
+  {};
+  /**
+   * UpSample the given Image
+   *
+   * @param input Input image.
+   * @param output The resulting interpolated output image.
+   */
+  template<typename eT>
+  void UpSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    // Get dimensions.
+    arma::Mat<eT> tempInput = input;
+    if (input.n_rows != inRowSize || input.n_cols != inColSize)
+      tempInput.reshape(inRowSize, inColSize);
+
+    output.set_size(outRowSize, outColSize);
+
+    if (output.n_rows > tempInput.n_rows)
+      scaleRow = inRowSize / outRowSize;
+    else
+      scaleRow = (inRowSize - 1) / outRowSize;
+
+    if (output.n_cols > tempInput.n_cols)
+      scaleCol = inColSize / outColSize;
+    else
+      scaleCol = (inColSize - 1) / outColSize;
+
+    for (size_t i = 0; i < outRowSize; i++)
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        r_origin = std::floor(i * scaleRow);
+        c_origin = std::floor(j * scaleCol);
+        GetCoff(i - r_origin, j - c_origin);
+        output = tempInput(r_origin, c_origin) * coff1 +
+                 tempInput(r_origin + 1, c_origin) * coff2 +
+                 tempInput(r_origin, c_origin + 1) * coff3 +
+                 tempInput(r_origin + 1, c_origin + 1) * coff4;
+      }
+  }
+
+  template<typename eT>
+  void DownSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    arma::Mat<eT> tempInput = input;
+    // Get dimensions.
+    if (input.n_rows != outRowSize || input.n_cols != outColSize)
+      tempInput.reshape(outRowSize, outColSize);
+
+    output.set_size(inRowSize, inColSize);
+
+    if (output.n_rows > tempInput.n_rows)
+      scaleRow = inRowSize / outRowSize;
+    else
+      scaleRow = (inRowSize - 1) / outRowSize;
+
+    if (output.n_cols > tempInput.n_cols)
+      scaleCol = inColSize / outColSize;
+    else
+      scaleCol = (inColSize - 1) / outColSize;
+
+    for (size_t i = 0; i < outRowSize; i++)
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        r_origin = std::floor(i * scaleRow);
+        c_origin = std::floor(j * scaleCol);
+        GetCoff(i - r_origin, j - c_origin);
```

Comment:
Looks like a typo. Maybe `Coeff`? I don't like the idea to change shared variables inside this function. In my opinion it is better to place the code of `GetCoff` here.